tools/drsnoop.py

```diff
@@ -0,0 +1,200 @@
+                    help="trace this UID only")
+parser.add_argument("-d", "--duration",
+                    help="total duration of trace in seconds")
+parser.add_argument("-n", "--name",
+                    type=ArgString,
+                    help="only print process names containing this name")
+parser.add_argument("--ebpf", action="store_true",
+                    help=argparse.SUPPRESS)
+args = parser.parse_args()
+debug = 0
+if args.duration:
+    args.duration = timedelta(seconds=int(args.duration))
+
+# load BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/sched.h>
+
+struct val_t {
+    u64 id;
+    u64 ts; // start time
+    char name[TASK_COMM_LEN];
+};
+
+struct data_t {
+    u64 id;
+    u32 uid;
+    u64 nr_reclaimed;
+    u64 delta;
+    u64 ts;    // end time
+    char name[TASK_COMM_LEN];
+};
+
+BPF_HASH(start, u64, struct val_t);
+BPF_PERF_OUTPUT(events);
+
+int trace_direct_reclaim_begin(void)
+{
+    struct val_t val = {};
+    u64 id = bpf_get_current_pid_tgid();
+    u32 pid = id >> 32; // PID is higher part
+    u32 tid = id;       // Cast and get the lower part
+    u32 uid = bpf_get_current_uid_gid();
+    u64 ts;
+
+    PID_TID_FILTER
+    UID_FILTER
+    if (bpf_get_current_comm(&val.name, sizeof(val.name)) == 0) {
+        val.id = id;
+        val.ts = bpf_ktime_get_ns();
+        start.update(&id, &val);
+    }
+    return 0;
+}
+
+struct reclaimed_args {
+    // from /sys/kernel/debug/tracing/events/vmscan/mm_vmscan_direct_reclaim_-
+    // end/format
+    u64 __unused__;
+    u64 nr_reclaimed;
+};
+
+int trace_direct_reclaim_end(struct reclaimed_args *args)
+{
+    u64 id = bpf_get_current_pid_tgid();
+    struct val_t *valp;
+    struct data_t data = {};
+    u64 ts = bpf_ktime_get_ns();
+
+    valp = start.lookup(&id);
+    if (valp == NULL) {
+        // missed entry
+        return 0;
+    }
+
+    data.delta = ts - valp->ts;
+    data.ts = ts / 1000;
+    data.id = valp->id;
+    data.uid = bpf_get_current_uid_gid();
+    bpf_probe_read(&data.name, sizeof(data.name), valp->name);
+    data.nr_reclaimed = args->nr_reclaimed;
+
+    events.perf_submit(args, &data, sizeof(data));
+    start.delete(&id);
+
+    return 0;
+}
+"""
+if args.tid:  # TID trumps PID
+    bpf_text = bpf_text.replace('PID_TID_FILTER',
+                                'if (tid != %s) { return 0; }' % args.tid)
+elif args.pid:
+    bpf_text = bpf_text.replace('PID_TID_FILTER',
+                                'if (pid != %s) { return 0; }' % args.pid)
+else:
+    bpf_text = bpf_text.replace('PID_TID_FILTER', '')
+if args.uid:
+    bpf_text = bpf_text.replace('UID_FILTER',
+                                'if (uid != %s) { return 0; }' % args.uid)
+else:
+    bpf_text = bpf_text.replace('UID_FILTER', '')
+if debug or args.ebpf:
+    print(bpf_text)
+    if args.ebpf:
+        exit()
+
+# initialize BPF
+b = BPF(text=bpf_text)
+b.attach_tracepoint(tp="vmscan:mm_vmscan_direct_reclaim_begin",
```

Comment:
If you use tracepoint macros, these explicit attach_tracepoint is not needed any more.

Reply:
refactored, PTAL.